sdk/dotnet/Pulumi.Automation/Events/EventLogWatcher.cs

```
@@ -0,0 +1,123 @@
+        private readonly Action<EngineEvent> _onEvent;
+        private readonly Task _pollingTask;
+
+        private const int _pollingIntervalMilliseconds = 100;
+
+        // We keep track of the length we have already read from the file
+        private long _previousLength = 0;
+
+        public string LogFile { get; }
+
+        internal EventLogWatcher(
+            string logFile
+            , Action<EngineEvent> onEvent
+            , CancellationToken externalCancellationToken)
+        {
+            LogFile = logFile;
+            _onEvent = onEvent;
+            _externalCancellationToken = externalCancellationToken;
+            _pollingTask = Task.Run(PollForEvents);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            _internalCancellationTokenSource.Cancel();
+            await _pollingTask.ConfigureAwait(false);
+            _internalCancellationTokenSource.Dispose();
+        }
+
+        private async Task PollForEvents()
+        {
+            var internalToken = _internalCancellationTokenSource.Token;
+            var externalToken = _externalCancellationToken;
+
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(internalToken, externalToken);
+
+            var linkedToken = linkedSource.Token;
+
+            while (!linkedToken.IsCancellationRequested)
+            {
+                try
+                {
+                    // NOTE: Waiting before reading so that if ReadEvents throws we will
+                    // wait before attempting it again
+                    await Task.Delay(_pollingIntervalMilliseconds, linkedToken).ConfigureAwait(false);
+                    ReadEvents();
+                }
+                catch (TaskCanceledException)
+                {
+                    break;
+                }
+                catch
+                {
+                    // We do our best effort to keep reading until the polling is cancelled
```

First location of ignored exceptions.